### PR TITLE
[ODS-4910] Year specific mode - xsd/ed-fi/files endpoint error

### PR DESCRIPTION
**Pull request by semalaiappan**

### CODE REVIEW

**Review by tbmeador-dlp — DISMISSED** [review]

**Comment by semalaiappan** [comment]
@tbmeador-dlp . Ready for review again . Thank you !

**Comment by vimayya** [comment]
@semalaiappan Why aren't we using the same pattern used by other features for route generation https://github.com/Ed-Fi-Alliance-OSS/Ed-Fi-ODS/blob/main/Application/EdFi.Ods.Features/Conventions/CompositesRouteConvention.cs#L81. If there is a reason for not using the same pattern, please include logic for handling InstanceYearSpecific mode.

**Comment by semalaiappan** [comment]
@vimayya  I added logic for handling InstanceYearSpecific mode in xmlMetadata controller .

File: Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs [hunk 1]
```diff
@@ -24,14 +26,17 @@ public class XsdMetadataController : ControllerBase
...
         public XsdMetadataController(ApiSettings settings,
             IXsdFileInformationProvider xsdFileInformationProvider,
-            IUrlHelper urlHelper)
+            IUrlHelper urlHelper,
+            ApiSettings apiSettings)
         {
             _xsdFileInformationProvider = xsdFileInformationProvider;
             _urlHelper = urlHelper;
             _isEnabled = settings.IsFeatureEnabled(ApiFeature.XsdMetadata.GetConfigKeyName());
+            _useProxyHeaders = apiSettings.UseReverseProxyHeaders.HasValue && apiSettings.UseReverseProxyHeaders.Value;
```
Comment (jamessmckay) [inline Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs:39]: should this really be a field? should you consider setting the settings as a field instead

File: Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs [hunk 2]
```diff
@@ -73,36 +78,57 @@ public IActionResult Get()
...
             }
 
             var results = schemaInformation.SchemaFiles
-                .Select(
-                    x => new Uri(
-                        $"{_urlHelper.ActionLink("Get", ControllerContext.ActionDescriptor.ControllerName)}/{schemaInformation.SchemaNameMap.UriSegment}/{x}"))
+                .Select(x=> GetMetadataAbsoluteUrl(x, schemaInformation.SchemaNameMap.UriSegment))
                 .ToList();
-
+
```
Comment (jamessmckay) [inline Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs:98]: remove extra whitespace please

File: Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs [hunk 3]
```diff
@@ -73,36 +78,57 @@ public IActionResult Get()
...
 
-                results.AddRange(
-                    coreInformation.SchemaFiles
-                        .Select(
-                            x => new Uri(
-                                $"{_urlHelper.ActionLink("Get", ControllerContext.ActionDescriptor.ControllerName)}/{coreInformation.SchemaNameMap.UriSegment}/{x}"))
-                );
+                results.AddRange(coreInformation.SchemaFiles
+                        .Select(x => GetMetadataAbsoluteUrl(x, coreInformation.SchemaNameMap.UriSegment))
+                        .ToList());
```
Comment (jamessmckay) [inline Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs:105]: why adding the list you are forcing an extra enumeration.

File: Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs [hunk 4]
```diff
@@ -73,36 +78,57 @@ public IActionResult Get()
...
+                {
+                    relativePath += $"/{request.InstanceIdFromRoute}";
+                }
+
+                if (request.SchoolYearFromRoute.HasValue)
+                {
+                    relativePath += $"/{request.SchoolYearFromRoute.Value}";
+                }
+
+                return $"{relativePath}/{ControllerContext.ActionDescriptor.ControllerName.Substring(0, 3)}/{uriSegment}/{schemaFile}";
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs:131]: fyi
```
bool isInstanceYearSpecific = _apiSettings.GetApiMode().Equals(ApiMode.InstanceYearSpecific);

bool isYearSpecific = _apiSettings.GetApiMode().Equals(ApiMode.YearSpecific)
                      || isInstanceYearSpecific;

bool useReverseProxyHeaders = _apiSettings.UseReverseProxyHeaders ?? false;
```

[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs:131]: I may make sense to create an ApiSettingsExtension class that moves this logic into it. This is from the Version Controller

**Review by jamessmckay — CHANGES_REQUESTED** [review]

**Comment by semalaiappan** [comment]
@jamessmckay   my  previous code of pattern was followed from OpenApiMetadataController file  , Now  I rework on this defect . based on your review comments . Note instance variable is not getting value in  VersionController ,so  I used _instanceIdContextProvider.GetInstanceId() for instance value in XsdMetadataController

File: Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs [hunk 5]
```diff
@@ -24,14 +26,17 @@ public class XsdMetadataController : ControllerBase
...
+        private readonly bool _useProxyHeaders;
 
         public XsdMetadataController(ApiSettings settings,
             IXsdFileInformationProvider xsdFileInformationProvider,
-            IUrlHelper urlHelper)
+            IUrlHelper urlHelper,
+            ApiSettings apiSettings)
         {
             _xsdFileInformationProvider = xsdFileInformationProvider;
             _urlHelper = urlHelper;
```
[the remark this answers is not in the document]
Comment (semalaiappan) [reply Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs]: I removed it . Thanks

File: Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs (hunk 1 above)
Comment (semalaiappan) [reply Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs]: I removed it . Thanks

File: Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs (hunk 2 above)
Comment (semalaiappan) [reply Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs]: I removed it . Thanks

File: Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs (hunk 3 above)
Comment (semalaiappan) [reply Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs]: I removed it . Thanks

**Comment by semalaiappan** [comment]
http://localhost:54746/metadata/26b76ab8-c8c3-43a9-9e5f-36356a658a46/2021/xsd/ed-fi/files

**Comment by semalaiappan** [comment]
@vimayya Question : YearSpecific is always currentyear ?

File: Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (semalaiappan) [reply Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs]: I implemented required logic here

**Comment by vimayya** [comment]
> @vimayya Question : YearSpecific is always currentyear ?

No, logic should work for any year.

**Comment by semalaiappan** [comment]
> > @vimayya Question : YearSpecific is always currentyear ?
> 
> No, logic should work for any year.

OK Thanks

**Comment by semalaiappan** [comment]
> > @vimayya Question : YearSpecific is always currentyear ?
> 
> No, logic should work for any year.

Then why in Version controller always display current year  https://github.com/Ed-Fi-Alliance-OSS/Ed-Fi-ODS/blob/main/Application/EdFi.Ods.Api/Controllers/VersionController.cs#L74

Application/EdFi.Ods.Api/Controllers/VersionController.cs:74

**Comment by blmeyers** [comment]
> > > @vimayya Question : YearSpecific is always currentyear ?
> > 
> > 
> > No, logic should work for any year.
> 
> Then why in Version controller always display current year https://github.com/Ed-Fi-Alliance-OSS/Ed-Fi-ODS/blob/main/Application/EdFi.Ods.Api/Controllers/VersionController.cs#L74
> 
> Application/EdFi.Ods.Api/Controllers/VersionController.cs:74

VersionController is accessed from the root, so it doesn't have route information about what year is being selected.  It defaults to current year, there's been discussion around swapping that to a replacement token similar to how instance id is handled.  In this scenario for xsds, a school year is available on the route, so the generated urls should reflect that value.

File: Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs [hunk 6]
```diff
@@ -88,24 +99,43 @@ public IActionResult GetFiles([FromRoute] string schema)
...
+            bool isInstanceYearSpecific = _apiSettings.GetApiMode().Equals(ApiMode.InstanceYearSpecific);
+
+            string instance = _instanceIdContextProvider.GetInstanceId();
+
+            bool isYearSpecific = _apiSettings.GetApiMode().Equals(ApiMode.YearSpecific)
+                                  || isInstanceYearSpecific;
+
+            bool useReverseProxyHeaders = _apiSettings.UseReverseProxyHeaders ?? false;
+
+            var currentYear = _systemDateProvider.GetDate().Year.ToString();
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs:127]: the controller should use the school year context provider

**Review by jamessmckay — CHANGES_REQUESTED** [review]

File: Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (semalaiappan) [reply Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs]: I fixed it . Thanks

File: Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs [hunk 7]
```diff
@@ -22,16 +26,23 @@ namespace EdFi.Ods.Features.XsdMetadata
...
+            IUrlHelper urlHelper,
+            ISystemDateProvider systemDateProvider,
+            IInstanceIdContextProvider instanceIdContextProvider,
+            ApiSettings apiSettings)
         {
             _xsdFileInformationProvider = xsdFileInformationProvider;
-            _urlHelper = urlHelper;
             _isEnabled = settings.IsFeatureEnabled(ApiFeature.XsdMetadata.GetConfigKeyName());
+            _apiSettings = apiSettings;
+            _systemDateProvider = systemDateProvider;
```
[the remark this answers is not in the document]
Comment (semalaiappan) [reply Application/EdFi.Ods.Features/XsdMetadata/XsdMetadataController.cs]: I fixed it . Thanks

**Review by jamessmckay — APPROVED** [review]

**Pull request merged**